csharp/src/Google.Protobuf/Reflection/CustomOptions.cs

```diff
@@ -205,98 +232,116 @@ internal CustomOptions(IDictionary<int, IExtensionValue> values)
+                        RepeatedExtensionValue<T> repeated = extensionValue as RepeatedExtensionValue<T>;
+                        if (repeated.GetValue().Count != 0)
                         {
-                            value = (T)typeInfo.GetDeclaredMethod(nameof(ExtensionValue<T>.GetValue)).Invoke(extensionValue, EmptyParameters);
+                            RepeatedField<T> repeatedField = repeated.GetValue();
+                            value = repeatedField[repeatedField.Count - 1];
                             return true;
                         }
                     }
-                    else if (type.GetGenericTypeDefinition() == typeof(RepeatedExtensionValue<>))
+                    else // and here we find explicit enum handling since T : Enum ! x is ExtensionValue<Enum>
                     {
-                        var typeInfo = type.GetTypeInfo();
-                        var typeArgs = typeInfo.GenericTypeArguments;
-                        if (typeArgs.Length == 1 && typeArgs[0].GetTypeInfo().IsEnum)
+                        var type = extensionValue.GetType();
+                        if (type.GetGenericTypeDefinition() == typeof(ExtensionValue<>))
                         {
-                            var values = (IList)typeInfo.GetDeclaredMethod(nameof(RepeatedExtensionValue<T>.GetValue)).Invoke(extensionValue, EmptyParameters);
-                            if (values.Count != 0)
+                            var typeInfo = type.GetTypeInfo();
+                            var typeArgs = typeInfo.GenericTypeArguments;
+                            if (typeArgs.Length == 1 && typeArgs[0].GetTypeInfo().IsEnum)
                             {
-                                value = (T)values[values.Count - 1];
+                                value = (T)typeInfo.GetDeclaredMethod(nameof(ExtensionValue<T>.GetValue)).Invoke(extensionValue, EmptyParameters);
                                 return true;
                             }
                         }
+                        else if (type.GetGenericTypeDefinition() == typeof(RepeatedExtensionValue<>))
+                        {
+                            var typeInfo = type.GetTypeInfo();
+                            var typeArgs = typeInfo.GenericTypeArguments;
+                            if (typeArgs.Length == 1 && typeArgs[0].GetTypeInfo().IsEnum)
+                            {
+                                var values = (IList)typeInfo.GetDeclaredMethod(nameof(RepeatedExtensionValue<T>.GetValue)).Invoke(extensionValue, EmptyParameters);
+                                if (values.Count != 0)
+                                {
+                                    value = (T)values[values.Count - 1];
+                                    return true;
+                                }
+                            }
+                        }
                     }
                 }
             }
 
+            UnknownField f;
```

Comment:
Without further context, from the code itself it's not clear at all why we would look at extensions first and then fall back to trying to read from unknown fields. There should really be a comment that explains why consulting why UnknownFieldSet is important.
Same for TryGetMessage.

Perhaps the refactoring I proposed would help to make this clearer.